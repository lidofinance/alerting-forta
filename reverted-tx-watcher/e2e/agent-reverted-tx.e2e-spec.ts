import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-reverted-tx e2e tests", () => {
  let runBlock: (
    blockHashOrNumber: string | number,
    initBlock?: number,
  ) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;
  let timeSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
    timeSpy = jest.spyOn(Date, "now");
    timeSpy.mockImplementation(() => new Date("2023-12-31"));
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(provideAgentPath("agent-reverted-tx")),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should find reverted tx in motions contract",
    async () => {
      const findings = await runTransaction(
        "0x99912053494d59ec0d0957f5e76e17f4e7d28baa6d5bbf886ef8b9356dc68425",
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );
});
