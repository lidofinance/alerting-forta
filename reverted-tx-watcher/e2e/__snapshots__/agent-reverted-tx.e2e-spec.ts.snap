// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-reverted-tx e2e tests should find reverted tx in motions contract 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "REVERTED-TX",
  "description": "Reverted TX to the EasyTrack contract. https://etherscan.io/tx/0x99912053494d59ec0d0957f5e76e17f4e7d28baa6d5bbf886ef8b9356dc68425",
  "labels": Array [],
  "metadata": Object {
    "sender": "0x07fe5f404778c27f4d3a0ab56dc59f8efdd32d96",
  },
  "name": "ðŸ¤” Reverted TX detected",
  "protocol": "ethereum",
  "severity": 1,
  "type": 2,
}
`;
